File: src/main/scala/kim/dencka/easy/collection/MajorityElement.scala
package kim.dencka.easy.collection

object MajorityElement extends App {

// #1
//  def majorityElement(nums: Array[Int]): Int = {
//    val half = nums.length / 2
//    val x = nums.groupBy(identity).view.mapValues(_.length).maxBy(_._2)
//    if (x._2 > half)  x._1
//    else 0
//  }

//  #2
//  def majorityElement(nums: Array[Int]): Int = {
//    val sorted = nums.sorted
//    sorted(sorted.length / 2)
//  }

// O(n) - time, O(1) - space
  def majorityElement(nums: Array[Int]): Int = {
    @annotation.tailrec
    def majorityElement(list: Array[Int], countMap: Map[Int, Int]): Int = {
      val current: Int = list.head
      val updatedCountMap: Map[Int, Int] = countMap + (current -> (countMap.getOrElse(current, 0) + 1))

      if (updatedCountMap(current) > nums.length / 2) current
      else majorityElement(list.tail, updatedCountMap)
    }
    majorityElement(nums, Map.empty)
  }


  println(majorityElement(Array(2,2,1,1,1,2,2)))
}

File: src/main/scala/kim/dencka/java/easy/collection/MajorityElement.java
package kim.dencka.java.easy.collection;

class MajorityElement {
    public static void main(String[] args) {
        int[] nums = {2, 2, 1, 1, 1, 2, 2};
        System.out.println(majorityElement(nums));
    }

    public static int majorityElement(int[] nums) {
        // cheating
//        nums[nums.length / 2]
        int cnt = 0;
        int maj = 0;

        for (int num : nums) {
            if (cnt == 0) maj = num;

            if (maj != num) cnt--;
            else cnt++;
        }
        return maj;
    }
}
